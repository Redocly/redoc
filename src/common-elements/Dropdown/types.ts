export interface DropdownOption {
  idx?: number;
  value: string;
  title?: string;
  serverUrl?: string;
  label?: string;
}

export interface DropdownProps {
  options: DropdownOption[];
  onChange: (option: DropdownOption) => void;
  ariaLabel?: string;
  className?: string;
  placeholder?: string;
  value?: string;
  dense?: boolean;
  fullWidth?: boolean;
  variant?: 'dark' | 'light';
}

export interface ArrowIconProps {
  className?: string;
  variant?: 'light' | 'dark';
  style?: React.CSSProperties;
}
