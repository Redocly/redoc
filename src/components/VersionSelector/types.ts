export interface ActiveVersionData {
  resourceVersion: string;
  apiVersion: string;
}
export interface VersionSelectorProps {
  resourceVersions: string[];
  active: ActiveVersionData;
  rootUrl: string;
  description?: string;
}

export interface OptionProps {
  focused: boolean;
  option: string;
  value: string;
  selected: boolean;
  onClick: (option: string) => void;
}

export interface ArrowIconProps {
  open: boolean;
}
