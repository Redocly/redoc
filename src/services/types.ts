import {
  OpenAPIOperation,
  OpenAPIParameter,
  OpenAPISchema,
  OpenAPIServer,
  OpenAPITag,
  Referenced,
} from '../types';
import { AppStore } from './AppStore';
import { GroupModel } from './models';
import { OperationModel } from './models/Operation';
import { RedocRawOptions } from './RedocNormalizedOptions';

export interface StoreState {
  menu: {
    activeItemIdx: number;
  };
  spec: {
    url?: string;
    data: any;
  };
  searchIndex: any;
  options: RedocRawOptions;
}

export interface LabelsConfig {
  enum: string;
  enumSingleValue: string;
  enumArray: string;
  default: string;
  deprecated: string;
  example: string;
  examples: string;
  recursive: string;
  arrayOf: string;
  webhook: string;
  const: string;
  noResultsFound: string;
  download: string;
  downloadSpecification: string;
  responses: string;
  callbackResponses: string;
  requestSamples: string;
  responseSamples: string;
}

export type LabelsConfigRaw = Partial<LabelsConfig>;

export interface MDXComponentMeta {
  component: React.ComponentType;
  propsSelector: (store?: AppStore) => any;
  props?: object;
}

export interface MarkdownHeading {
  id: string;
  name: string;
  level: number;
  items?: MarkdownHeading[];
  description?: string;
}

export type ContentItemModel = GroupModel | OperationModel;

export type TagInfo = OpenAPITag & {
  operations: ExtendedOpenAPIOperation[];
  used?: boolean;
};

export type ExtendedOpenAPIOperation = {
  pointer: string;
  pathName: string;
  httpVerb: string;
  pathParameters: Array<Referenced<OpenAPIParameter>>;
  pathServers: Array<OpenAPIServer> | undefined;
  isWebhook: boolean;
} & OpenAPIOperation;

export type TagsInfoMap = Record<string, TagInfo>;

export interface TagGroup {
  name: string;
  tags: string[];
}

export type MenuItemGroupType = 'group' | 'tag' | 'section';
export type MenuItemType = MenuItemGroupType | 'operation';

export interface IMenuItem {
  id: string;
  absoluteIdx?: number;
  name: string;
  sidebarLabel: string;
  description?: string;
  depth: number;
  active: boolean;
  expanded: boolean;
  items: IMenuItem[];
  parent?: IMenuItem;
  deprecated?: boolean;
  type: MenuItemType;

  deactivate(): void;
  activate(): void;

  collapse(): void;
  expand(): void;
}

export interface SearchDocument {
  title: string;
  description: string;
  id: string;
}

export interface SearchResult<T = string> {
  meta: T;
  score: number;
}

export enum SideNavStyleEnum {
  SummaryOnly = 'summary-only',
  PathOnly = 'path-only',
  IdOnly = 'id-only',
}

export type MergedOpenAPISchema = OpenAPISchema & {
  'x-refsStack'?: string[];
  'x-parentRefs'?: string[];
  'x-circular-ref'?: boolean;
};
