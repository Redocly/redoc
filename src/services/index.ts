export * from './AppStore';
export * from './OpenAPIParser';
export * from './MarkdownRenderer';
export * from './MenuStore';
export * from './ScrollService';
export * from './SpecStore';
export * from './ClipboardService';
export * from './HistoryService';
export * from './models';
export * from './RedocNormalizedOptions';
export * from './MenuBuilder';
export * from './SearchStore';
export * from './MarkerService';
export * from './types';
