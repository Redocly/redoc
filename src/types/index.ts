export * from './open-api';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
