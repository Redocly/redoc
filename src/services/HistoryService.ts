import { bind, debounce } from 'decko';
import { EventEmitter } from 'eventemitter3';
import { IS_BROWSER } from '../utils/';
import { RedocNormalizedOptions } from './RedocNormalizedOptions';

const EVENT = 'hashchange';

export class HistoryService {
  private _emiter;
  private options: RedocNormalizedOptions;

  constructor(options: RedocNormalizedOptions) {
    this.options = options;
    this._emiter = new EventEmitter();
    this.bind();
  }

  get currentId(): string {
    if (IS_BROWSER) {
      if (this.shouldQueryParamNavigationBeUsed()) {
        // When the window.location.hash is not empty this means that we have clicked on
        // router that's for example stored in the description via markdown
        if (window.location.hash == '') {
          return this.getQueryParams(window.location.search);
        } else {
          return decodeURIComponent(window.location.hash.substring(1));
        }
      } else {
        return decodeURIComponent(window.location.hash.substring(1));
      }
    }
    return '';
  }

  linkForId(id: string) {
    if (!id) {
      return '';
    }
    if (this.shouldQueryParamNavigationBeUsed()) {
      return this.getFullUrl(id);
    } else {
      return '#' + id;
    }
  }

  subscribe(cb): () => void {
    const emmiter = this._emiter.addListener(EVENT, cb);
    return () => emmiter.removeListener(EVENT, cb);
  }

  emit = () => {
    this._emiter.emit(EVENT, this.currentId);
  };

  bind() {
    if (IS_BROWSER) {
      window.addEventListener('hashchange', this.emit, false);
    }
  }

  dispose() {
    if (IS_BROWSER) {
      window.removeEventListener('hashchange', this.emit);
    }
  }

  @bind
  @debounce
  replace(id: string | null, rewriteHistory: boolean = false) {
    if (!IS_BROWSER) {
      return;
    }

    // If there currentId and the ID are equal but there is still
    // a hash left when using query param navigation
    // that means that the URL hasn't been overridden
    if (
      id == null ||
      (id === this.currentId && this.checkIfThereIsHashLeftWhenQueryParamNavigationShouldBeUsed())
    ) {
      return;
    }
    if (rewriteHistory) {
      if (this.shouldQueryParamNavigationBeUsed()) {
        window.history.replaceState(null, '', this.getFullUrl(id));
      } else {
        window.history.replaceState(
          null,
          '',
          window.location.href.split('#')[0] + this.linkForId(id),
        );
      }

      return;
    }
    if (this.shouldQueryParamNavigationBeUsed()) {
      window.history.pushState(null, '', this.getFullUrl(id));
    } else {
      window.history.pushState(null, '', window.location.href.split('#')[0] + this.linkForId(id));
    }
    this.emit();
  }

  private shouldQueryParamNavigationBeUsed(): boolean {
    return this.options?.userQueryParamToNavigate;
  }

  private getQueryParams(search: string): string {
    const queryParams = new URLSearchParams(search);
    if (search != null) {
      return queryParams.get('redoc') != null ? (queryParams.get('redoc') as string) : '';
    }
    return '';
  }

  private getFullUrl(id: string): string {
    const url = this.getUrl();
    // Override the hash, so it's removed when using query param navigation
    url.hash = '';
    url.searchParams.set('redoc', id);
    return url.toString();
  }

  private getUrl(): URL {
    return new URL(window.location.href);
  }

  private checkIfThereIsHashLeftWhenQueryParamNavigationShouldBeUsed(): boolean {
    return !(this.shouldQueryParamNavigationBeUsed() && window.location.hash != '');
  }
}
