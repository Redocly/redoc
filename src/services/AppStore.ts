import { Lambda, observe } from 'mobx';

import { OpenAPISpec } from '../types';
import { loadAndBundleSpec } from '../utils/loadAndBundleSpec';
import { history } from './HistoryService';
import { MarkerService } from './MarkerService';
import { MenuStore } from './MenuStore';
import { SpecStore } from './models';
import { RedocNormalizedOptions, RedocRawOptions } from './RedocNormalizedOptions';
import { ScrollService } from './ScrollService';
import { SearchStore } from './SearchStore';

import { SchemaDefinition } from '../components/SchemaDefinition/SchemaDefinition';
import { SecurityDefs } from '../components/SecuritySchemes/SecuritySchemes';
import {
  SCHEMA_DEFINITION_JSX_NAME,
  SECURITY_DEFINITIONS_COMPONENT_NAME,
  SECURITY_DEFINITIONS_JSX_NAME,
} from '../utils/openapi';

import { IS_BROWSER } from '../utils';

export interface StoreState {
  menu: {
    activeItemIdx: number;
  };
  spec: {
    url?: string;
    data: any;
  };
  searchIndex: any;
  options: RedocRawOptions;
}

export async function createStore(
  spec: object,
  specUrl: string | undefined,
  options: RedocRawOptions = {},
) {
  const resolvedSpec = await loadAndBundleSpec(spec || specUrl);
  return new AppStore(resolvedSpec, specUrl, options);
}

export class AppStore {
  /**
   * deserialize store
   * **SUPER HACKY AND NOT OPTIMAL IMPLEMENTATION**
   */
  // TODO:
  static fromJS(state: StoreState): AppStore {
    const inst = new AppStore(state.spec.data, state.spec.url, state.options, false);
    inst.menu.activeItemIdx = state.menu.activeItemIdx || 0;
    inst.menu.activate(inst.menu.flatItems[inst.menu.activeItemIdx]);
    if (!inst.options.disableSearch) {
      inst.search!.load(state.searchIndex);
    }
    return inst;
  }

  menu: MenuStore;
  spec: SpecStore;
  rawOptions: RedocRawOptions;
  options: RedocNormalizedOptions;
  search?: SearchStore<string>;
  marker = new MarkerService();

  private scroll: ScrollService;
  private disposer: Lambda | null = null;

  constructor(
    spec: OpenAPISpec,
    specUrl?: string,
    options: RedocRawOptions = {},
    createSearchIndex: boolean = true,
  ) {
    this.rawOptions = options;
    this.options = new RedocNormalizedOptions(options, DEFAULT_OPTIONS);
    this.scroll = new ScrollService(this.options);

    // update position statically based on hash (in case of SSR)
    MenuStore.updateOnHistory(history.currentId, this.scroll);

    this.spec = new SpecStore(spec, specUrl, this.options);
    this.menu = new MenuStore(this.spec, this.scroll, history);

    if (!this.options.disableSearch) {
      this.search = new SearchStore();
      if (createSearchIndex) {
        this.search.indexItems(this.menu.items);
      }

      this.disposer = observe(this.menu, 'activeItemIdx', change => {
        this.updateMarkOnMenu(change.newValue as number);
      });
    }
  }

  onDidMount() {
    this.menu.updateOnHistory();
    this.updateMarkOnMenu(this.menu.activeItemIdx);
  }

  dispose() {
    this.scroll.dispose();
    this.menu.dispose();
    if (this.search) {
      this.search.dispose();
    }
    if (this.disposer != null) {
      this.disposer();
    }
  }

  /**
   * serializes store
   * **SUPER HACKY AND NOT OPTIMAL IMPLEMENTATION**
   */
  // TODO: improve
  async toJS(): Promise<StoreState> {
    return {
      menu: {
        activeItemIdx: this.menu.activeItemIdx,
      },
      spec: {
        url: this.spec.parser.specUrl,
        data: this.spec.parser.spec,
      },
      searchIndex: this.search ? await this.search.toJS() : undefined,
      options: this.rawOptions,
    };
  }

  private updateMarkOnMenu(idx: number) {
    const start = Math.max(0, idx);
    const end = Math.min(this.menu.flatItems.length, start + 5);

    const elements: Element[] = [];
    for (let i = start; i < end; i++) {
      const elem = this.menu.getElementAt(i);
      if (!elem) {
        continue;
      }
      elements.push(elem);
    }

    if (idx === -1 && IS_BROWSER) {
      const $description = document.querySelector('[data-role="redoc-description"]');
      const $summary = document.querySelector('[data-role="redoc-summary"]');

      if ($description) elements.push($description);
      if ($summary) elements.push($summary);
    }

    this.marker.addOnly(elements);
    this.marker.mark();
  }
}

const DEFAULT_OPTIONS: RedocRawOptions = {
  allowedMdComponents: {
    [SECURITY_DEFINITIONS_COMPONENT_NAME]: {
      component: SecurityDefs,
      propsSelector: (store: AppStore) => ({
        securitySchemes: store.spec.securitySchemes,
      }),
    },
    [SECURITY_DEFINITIONS_JSX_NAME]: {
      component: SecurityDefs,
      propsSelector: (store: AppStore) => ({
        securitySchemes: store.spec.securitySchemes,
      }),
    },
    [SCHEMA_DEFINITION_JSX_NAME]: {
      component: SchemaDefinition,
      propsSelector: (store: AppStore) => ({
        parser: store.spec.parser,
        options: store.options,
      }),
    },
  },
};
