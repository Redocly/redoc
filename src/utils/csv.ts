import * as Sampler from 'openapi-sampler';
import { OpenAPIExample, OpenAPISchema } from '../types';
import { Example } from '../types/example';
import { MergedOpenAPISchema, OpenAPIParser } from '../services';

const MAX_ITEM_DEPTH = 1;

interface CsvExampleProps {
  parser: OpenAPIParser;
  schema: OpenAPISchema;
  sample: OpenAPIExample;
  samplerOptions: object;
}

const hasSameHeaders = (headers: string, sample: OpenAPIExample) =>
  Object.keys(sample).every(key => headers.includes(key));

const getCsvRows = (sample: OpenAPIExample): string => {
  const headers = Object.keys(sample?.[0] ?? sample).join(',');
  // Ensure the schema has deterministic headers
  const hasValidHeaders = (Array.isArray(sample) ? sample : [sample]).every(row =>
    hasSameHeaders(headers, row),
  );
  if (!hasValidHeaders) return '';

  let values: string;

  if (Array.isArray(sample)) {
    values = sample.map(Object.values).join('\n');
  } else {
    values = Object.values(sample).join(',');
  }
  return headers + '\n' + values;
};

const cleanUpExamples = (examples: Example[]): Example[] =>
  examples.filter(({ exampleValue }) => exampleValue);

export const generateCsvExample = ({
  parser,
  schema,
  sample,
  samplerOptions,
}: CsvExampleProps): Example[] => {
  let examples: Example[] = [];
  let depthCount = 0;
  let exampleCount = 1;
  const isValidSample = (Array.isArray(sample) ? sample : [sample]).every(sampleItem =>
    Object.values(sampleItem).every(
      value => typeof value !== 'object' && typeof value !== 'undefined',
    ),
  );

  const processSamplesWithSchema = (subSchema: OpenAPISchema) => {
    if (!subSchema) {
      return;
    }

    const subItems = subSchema.items as OpenAPISchema;
    if (subSchema.type === 'array' && subItems && depthCount < MAX_ITEM_DEPTH) {
      depthCount++;
      processSamplesWithSchema(subItems);
    }
    const metadata = {
      exampleDescription: subSchema.description || schema.description || '',
      exampleSummary: subSchema.title || schema.title || 'Example CSV',
    };
    if (subSchema.allOf) {
      const resolved: OpenAPISchema = {
        ...schema,
        items: parser.deref(subSchema.allOf as MergedOpenAPISchema).resolved,
      };
      const sampleData = Sampler.sample(
        resolved as any,
        samplerOptions,
        parser.spec,
      ) as OpenAPIExample;

      const csvRows = getCsvRows(sampleData);
      examples.push({
        exampleId: `Example ${exampleCount++}`,
        exampleValue: csvRows,
        ...metadata,
      });
    } else if (subSchema.oneOf) {
      const oneOfExamples = subSchema.oneOf.map(oneOfSchema => {
        const { resolved } = parser.deref(oneOfSchema as MergedOpenAPISchema);
        const sampleData = Sampler.sample(
          resolved as any,
          samplerOptions,
          parser.spec,
        ) as OpenAPIExample;
        const csvRows = getCsvRows(sampleData);
        const currentMetadata = {
          exampleDescription: oneOfSchema.description || metadata.exampleDescription,
          exampleSummary: oneOfSchema.title || metadata.exampleSummary,
        };

        return {
          exampleId: `Example ${exampleCount++}`,
          exampleValue: csvRows,
          ...currentMetadata,
        };
      });
      examples = [...examples, ...oneOfExamples];
    } else if (subSchema.$ref) {
      const csvRows = getCsvRows(sample);
      examples.push({
        exampleId: `Example ${exampleCount++}`,
        exampleValue: csvRows,
        ...metadata,
      });
    }
  };

  if (isValidSample) {
    processSamplesWithSchema(schema);
  }

  return cleanUpExamples(examples);
};
