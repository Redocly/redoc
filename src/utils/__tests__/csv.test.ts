import { generateCsvExample } from '../csv';
import { OpenAPIParser, RedocNormalizedOptions } from '../../services';
import { OpenAPIExample } from '../../types';

const opts = new RedocNormalizedOptions({});
const samplerOptions = {};

describe('generateCsvExample', () => {
  const spec = require('./fixtures/csv-compatible-schema.json');
  let parser;

  it('generates a csv example for an array of items using allOf', () => {
    parser = new OpenAPIParser(spec, undefined, opts);
    const sample = [
      {
        Competitor: 'google.com',
        '2023-02-02': 0.356263,
        '2023-02-03': 0.1251661,
      },
      {
        Competitor: 'facebook.com',
        '2023-02-02': 0.74324,
        '2023-02-03': 0.73542,
      },
    ] as unknown as OpenAPIExample;

    const examples = generateCsvExample({
      parser,
      schema: spec.components.schemas.test,
      sample,
      samplerOptions,
    });

    expect(examples).toMatchSnapshot();
  });

  it('generates a csv example for an array of items using oneOf', () => {
    parser = new OpenAPIParser(spec, undefined, opts);
    const sample = [
      {
        Competitor: 'facebook.com',
        '2023-02-02': 0.74324,
        '2023-02-03': 0.73542,
      },
    ] as unknown as OpenAPIExample;

    const examples = generateCsvExample({
      parser,
      schema: spec.components.schemas.test2,
      sample,
      samplerOptions,
    });

    expect(examples).toMatchSnapshot();
  });

  it('generates a csv example for an array of items using $ref', () => {
    parser = new OpenAPIParser(spec, undefined, opts);
    const sample = [
      {
        Competitor: 'google.com',
        '2023-02-02': 0.356263,
        '2023-02-03': 0.1251661,
      },
    ] as unknown as OpenAPIExample;

    const examples = generateCsvExample({
      parser,
      schema: spec.components.schemas.test3,
      sample,
      samplerOptions,
    });

    expect(examples).toMatchSnapshot();
  });

  it.each([
    {
      prop: [],
    },
    {
      prop2: {},
    },
    {
      prop3: null,
    },
    {
      prop4: undefined,
    },
  ] as unknown[] as OpenAPIExample[])(
    'should not generate a csv example',
    (sample: OpenAPIExample) => {
      parser = new OpenAPIParser(spec, undefined, opts);
      const examples = generateCsvExample({
        parser,
        schema: spec.components.schemas.test,
        sample,
        samplerOptions,
      });

      expect(examples.length).toEqual(0);
    },
  );
});
