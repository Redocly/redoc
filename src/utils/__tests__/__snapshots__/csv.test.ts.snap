// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCsvExample generates a csv example for an array of items using $ref 1`] = `
Array [
  Object {
    "exampleDescription": "",
    "exampleId": "Example 1",
    "exampleSummary": "Example CSV",
    "exampleValue": "Competitor,2023-02-02,2023-02-03
google.com,0.356263,0.1251661",
  },
]
`;

exports[`generateCsvExample generates a csv example for an array of items using allOf 1`] = `
Array [
  Object {
    "exampleDescription": "",
    "exampleId": "Example 1",
    "exampleSummary": "Example CSV",
    "exampleValue": "Competitor,2023-02-02,2023-02-03
google.com,0.356263,0.1251661
facebook.com,0.74324,0.73542",
  },
]
`;

exports[`generateCsvExample generates a csv example for an array of items using oneOf 1`] = `
Array [
  Object {
    "exampleDescription": "",
    "exampleId": "Example 1",
    "exampleSummary": "Example CSV",
    "exampleValue": "Competitor,2023-02-02,2023-02-03
google.com,0.356263,0.1251661",
  },
  Object {
    "exampleDescription": "",
    "exampleId": "Example 2",
    "exampleSummary": "Example CSV",
    "exampleValue": "Competitor,2023-02-02,2023-02-03
facebook.com,0.74324,0.73542",
  },
]
`;
